pkg/cli/gettoken/oidc/client.go

```diff
@@ -0,0 +1,227 @@
+package oidc
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	gooidc "github.com/coreos/go-oidc/v3/oidc"
+	"github.com/int128/oauth2cli"
+	"golang.org/x/oauth2"
+)
+
+// Authenticator defines the basic functionality to support Auth Code Flow
+// and Token Refresh Grant Flow.
+type Authenticator interface {
+	GetTokenByAuthCode(ctx context.Context, callbackAddress string, localServerReadyChan chan<- string) (string, string, time.Time, error)
+	Refresh(ctx context.Context, refreshToken string) (string, string, time.Time, error)
+	VerifyToken(ctx context.Context, token *oauth2.Token, nonce string) (string, string, time.Time, error)
+}
+
+// client manages the request/response flow between
+// the client and the auth server. It has a capability
+// to modify the http client with the given certificate
+// authority and pre-defined proxy settings in addition to
+// timeout value that can be customizable by the user.
+type client struct {
+	httpClient      *http.Client
+	provider        *gooidc.Provider
+	oauth2Config    oauth2.Config
+	isPKCESupported bool
+}
+
+func (c *client) GetTokenByAuthCode(ctx context.Context, callbackAddress string, localServerReadyChan chan<- string) (string, string, time.Time, error) {
+	bs := make([]byte, 32)
+	if err := binary.Read(rand.Reader, binary.LittleEndian, bs); err != nil {
+		return "", "", time.Time{}, fmt.Errorf("could not generate a state: %w", err)
+	}
+	state := base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(bs)
+
+	bn := make([]byte, 32)
+	if err := binary.Read(rand.Reader, binary.LittleEndian, bn); err != nil {
+		return "", "", time.Time{}, fmt.Errorf("could not generate a nonce: %w", err)
+	}
+	nonce := base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(bn)
+
+	var p string
+	if c.isPKCESupported {
+		p = oauth2.GenerateVerifier()
+	}
+
+	if c.httpClient != nil {
+		ctx = context.WithValue(ctx, oauth2.HTTPClient, c.httpClient)
+	}
+
+	config := oauth2cli.Config{
+		OAuth2Config:           c.oauth2Config,
+		State:                  state,
+		AuthCodeOptions:        authorizationRequestOptions(nonce, p),
+		TokenRequestOptions:    tokenRequestOptions(p),
+		LocalServerBindAddress: []string{callbackAddress},
+		LocalServerReadyChan:   localServerReadyChan,
+	}
+	token, err := oauth2cli.GetToken(ctx, config)
+	if err != nil {
+		return "", "", time.Time{}, fmt.Errorf("oauth2 error: %w", err)
+	}
+	idToken, refreshToken, expiry, err := c.VerifyToken(ctx, token, nonce)
+	return idToken, refreshToken, expiry, err
+}
+
+func (c *client) Refresh(ctx context.Context, refreshToken string) (string, string, time.Time, error) {
+	if c.httpClient != nil {
+		ctx = context.WithValue(ctx, oauth2.HTTPClient, c.httpClient)
+	}
+
+	currentToken := &oauth2.Token{
+		Expiry:       time.Now(),
```

Comment:
Isn't now a little bit early? Add 5 minutes? Comment otherwise.
Do we actually have to set it?

Reply:
Yes, adding a 5 * minutes representing the authentication timeout is a good idea. Added. 

> Do we actually have to set it?

According to here it wouldn't hurt https://github.com/openshift/oc/blob/076e3a363ea8f86522ae8ad0e16d70dd7a639d1f/vendor/golang.org/x/oauth2/token.go#L45-L50